File: rxbinding/src/main/java/com/jakewharton/rxbinding2/widget/RatingBarRatingChangeEventObservable.java
package com.jakewharton.rxbinding2.widget;

import android.widget.RatingBar;
import android.widget.RatingBar.OnRatingBarChangeListener;
import com.jakewharton.rxbinding2.InitialValueObservable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static com.jakewharton.rxbinding2.internal.Preconditions.checkMainThread;

final class RatingBarRatingChangeEventObservable
    extends InitialValueObservable<RatingBarChangeEvent> {
  private final RatingBar view;

  RatingBarRatingChangeEventObservable(RatingBar view) {
    this.view = view;
  }

  @Override protected void subscribeListener(Observer<? super RatingBarChangeEvent> observer) {
    if (!checkMainThread(observer)) {
      return;
    }
    Listener listener = new Listener(view, observer);
    view.setOnRatingBarChangeListener(listener);
    observer.onSubscribe(listener);
  }

  @Override protected RatingBarChangeEvent getInitialValue() {
    return RatingBarChangeEvent.create(view, view.getRating(), false);
  }

  static final class Listener extends MainThreadDisposable implements OnRatingBarChangeListener {
    private final RatingBar view;
    private final Observer<? super RatingBarChangeEvent> observer;

    Listener(RatingBar view, Observer<? super RatingBarChangeEvent> observer) {
      this.view = view;
      this.observer = observer;
    }

    @Override public void onRatingChanged(RatingBar ratingBar, float rating, boolean fromUser) {
      if (!isDisposed()) {
        observer.onNext(RatingBarChangeEvent.create(ratingBar, rating, fromUser));
      }
    }

    @Override protected void onDispose() {
      view.setOnRatingBarChangeListener(null);
    }
  }
}

File: rxbinding/src/main/java/com/jakewharton/rxbinding4/widget/RatingBarRatingChangeEventObservable.kt
@file:JvmName("RxRatingBar")
@file:JvmMultifileClass

package com.jakewharton.rxbinding4.widget

import android.widget.RatingBar
import android.widget.RatingBar.OnRatingBarChangeListener
import androidx.annotation.CheckResult
import com.jakewharton.rxbinding4.InitialValueObservable
import io.reactivex.rxjava3.core.Observer
import io.reactivex.rxjava3.android.MainThreadDisposable

import com.jakewharton.rxbinding4.internal.checkMainThread

/**
 * Create an observable of the rating change events on `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 *
 * *Note:* A value will be emitted immediately on subscribe.
 */
@CheckResult
fun RatingBar.ratingChangeEvents(): InitialValueObservable<RatingBarChangeEvent> {
  return RatingBarRatingChangeEventObservable(this)
}

data class RatingBarChangeEvent(
  /** The view from which this event occurred.  */
  val view: RatingBar,
  val rating: Float,
  val fromUser: Boolean
)

private class RatingBarRatingChangeEventObservable(
  private val view: RatingBar
) : InitialValueObservable<RatingBarChangeEvent>() {

  override fun subscribeListener(observer: Observer<in RatingBarChangeEvent>) {
    if (!checkMainThread(observer)) {
      return
    }
    val listener = Listener(view, observer)
    view.onRatingBarChangeListener = listener
    observer.onSubscribe(listener)
  }

  override val initialValue get() = RatingBarChangeEvent(view, view.rating, false)

  private class Listener(
    private val view: RatingBar,
    private val observer: Observer<in RatingBarChangeEvent>
  ) : MainThreadDisposable(), OnRatingBarChangeListener {

    override fun onRatingChanged(ratingBar: RatingBar, rating: Float, fromUser: Boolean) {
      if (!isDisposed) {
        observer.onNext(RatingBarChangeEvent(ratingBar, rating, fromUser))
      }
    }

    override fun onDispose() {
      view.onRatingBarChangeListener = null
    }
  }
}
